### ISSUE TITLE
Aliases for column names

### ISSUE THREAD

**Issue by bbirand** [pandas-dev/pandas#11723]
When I work with Pandas DataFrames, I prefer to keep the full column names for clarity. So when I print out the `head`, or use describe, I get a meaningful table. However, this also means I have column names like "Time of Sale" that become annoying to type out. 

A nice compromise seems like it would be to have short "aliases" for column names. For instance, I can define the `tos` average for the above, perhaps like so:

```
df = pd.read_csv(...)
df.set_alias({'Time of Sale' : 'tos'})
```

Then, the `__get_attribute__` method can look up aliases in addition to column names, so I can refer to that column simply as `df.tos`. But for all other purposes, the columns name is still the descriptive full name.

Would this make sense?

**Comment by jreback** [comment]
related to #10349 

I suppose this is possible. This would be fairly easy to implement, but would require a good number of test cases to ensure its propogating correctly (e.g. this is analagous to the ``name`` attribute for ``Indexes`` in that it propogates when appropriate).

Further would require an audit of the indexing code for it to be a synonymous application (e.g. you can use the alias where you could use the actual label).

So while this is interesting, it would require a pull-request from the community to jump start it.

**Comment by bbirand** [comment]
I'll have a go at this when I get a chance. It also occurred to me that these aliases may be useful when dealing with `DataFrame.query()` methods. Based on my trials, this function does not work when there are spaces on the column names (please correct me if I'm wrong, I couldn't get them to work).

**Comment by jreback** [comment]
no ``.query`` processes strings so you cannot use strings, this is noted in the documentation.

**Comment by shoyer** [comment]
I'm not a big fan of including this feature in pandas itself, because it would make the pandas data model significantly more complex. Maybe this could be implemented in some sort of add-on package that wraps pandas DataFrames? Another option would be a [DataFrame subclass](http://pandas.pydata.org/pandas-docs/version/0.17.1/internals.html#subclassing-pandas-data-structures).

**Comment by ijstokes** [comment]
There are certainly risks that could be introduced from adding aliasing, but wouldn't a straightforward strategy be to augment the logic in `get_attribute()` that, presumably, already does some form of this.  So if an `alias` dictionary existed on the `DataFrame` then it would try again provided the requested attribute (not found using "the usual mechanism") had a key entry in the `alias` dictionary.  E.g.

```python
# 1. works today:
df['Time of Sale']

# 2. fails today:
df.time_of_sale

# 3. could work in the future:
df.alias = dict(time_of_sale='Time of Sale')
df.time_of_sale
```

Or maybe I misunderstand and 2. is already possible today.  If so, could someone point me in the right direction toward documentation?  I too would find this quite useful.

**Comment by bbirand** [comment]
> Or maybe I misunderstand and 2. is already possible today. If so, could someone point me in the right direction toward documentation? I too would find this quite useful.
 
In order to do 2., you would have to rename the column, possibly using http://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.rename.html

And then, when you'd like to print or plot it, you'd rename it back to the original version.

I too think this would still be a good addition for interactive work. To make things even more interesting, I would alias "Time of Sale" to "tos", so I can work with the data as `df.tos`, but then see the full name when plotted.

**Comment by Plumbob** [comment]
I too would like to have more attractive column names when displaying Pandas DataFrames.  Would a viable compromise be to apply the aliases only when converting DataFrames to display formats?
To use bbirand's example, the 'tos' would be used internally, but the alias would be used only to convert the internal 'tos' column name to a more public-friendly 'Time of Sale' column name in __str__() output, for example.

**Comment by KeithWM** [comment]
I'd also like to see such a feature. For me the favourite use case would be to have nice, legible axes labels (with units) in seaborn plots. I know one can manually set the axis labels, but I find this error prone, too verbose and it leads to code duplication.

If you ask me, the easier way would be to keep the _current_ name in the role of an alias as @bbirand proposes, and to add some other field for a longer name, which can default to the "normal" name if none is explicitly given.

**Comment by ajeet2808** [comment]
Any update on this feature?

**Comment by luisfelipe18** [comment]
we need and equivalent for "SELECT max(column1)*0.25+ 0.44*sum(column2) as 'calculated_column' from TABLE group by column3,column5"

**Comment by obarak** [comment]
> we need and equivalent for `"SELECT max(column1)*0.25+ 0.44*sum(column2) as 'calculated_column' from TABLE group by column3,column5"`

@luisfelipe18 - Actually, for aggregation you already have aliasing in Pandas, see [here](https://deanla.com/pandas_named_agg.html#Replacing-the-as-keyword-from-SQL) (I'd recommend reading through the entire post).

The current issue refers to aliasing *existing columns*, regardless of aggregation.

**Comment by luisfelipe18** [comment]
@obarak yes, i'm aware of that. My point is that in aggregation you can apply a custom function to only one column, but what if I need to do in a pythonic way something like `df.groupby(column).agg(desired_name = pd.AggName(column_NName = "col_name", aggfunc = lambda x: sum(x) + 0.45*sum(column_z)))` ? 
the only way is adding to `agg` the `sum()` for column  `column_z` and then do the operation `df["column_NName"] = df.col_name + 0.45*df.column_z`

**Comment by TomAugspurger** [comment]
IMO, we shouldn't use this in pandas itself. Indexing is complicated enough without aliases.

We'd be better served by adopting / defining a convention (similar to how xarray uses [CF conventions](http://cfconventions.org/)) for mapping column names to descriptive names. These could be stored in the `DataFrame.attrs` dict which (should) propagate through operations. Then downstream libraries (e.g. plotting libraries, libraries for generating tables for presentation) can use the descriptive names.

**Comment by adavidzh** [comment]
I'd like to echo @KeithWM's point about there being a "long form" for a column's contents, i.e. something that `seaborn` can use in axis labels. This would not necessarily be an alias, but rather a human-readable form with full description (e.g. involving LaTeX) and units. This sort of thing comes up over and over again in making scientific plots; I found this thread because I want a column named `engine_data_total_throughput` and would like the axis label to be `$\sum$ data throughput [Gb/s]` without having to specify it over and over again when plotting.

I understand that this "long form" is not a general scheme for creating aliases (that is a many-to-one correspondence) and it could make sense to understand what is the main use case and, perhaps, have a new thread.

**Comment by jbrockmendel** [comment]
I implemented something like this for a personal project years ago, so I understand the reasons why this is appealing.  But it introduces a ton of complexity.  Off the top of my head

1) Every path in get_loc/get_indexer needs to check for aliases
2) Does `index == alias1` work like `index == canonical1`?
3) These aren't going to round-trip nicely through reset_index().set_index()
4) What if `alias1` happens to match `canonical2`?  What if `alias3 == alias4`?

i.e. getting the very simple use case the OP wants to work isn't that hard.  Getting it to work and be consistent is a giant mess.

**Comment by jbrockmendel** [comment]
Discussed on today's dev call, decided not to move forward here.
